File: math/add_strings.rb
# @param {String} num1
# @param {String} num2
# @return {String}
def add_strings(num1, num2)
    i, j = num1.size - 1, num2.size - 1
    result, carry = '', 0
    
    while i >= 0 || j >= 0
        d1 = i >= 0 ? num1[i].to_i : 0
        d2 = j >= 0 ? num2[j].to_i : 0
        
        carry, d = (d1 + d2 + carry).divmod(10)
        result = d.to_s + result
        i -= 1; j -= 1
    end

    carry == 0 ? result : carry.to_s + result
end

# 415. Add Strings
# https://leetcode.com/problems/add-strings/

# Time: O(n)
# Space: O(1)

require 'test/unit'
extend Test::Unit::Assertions

assert_equal(add_strings('21', '12'), 33)
assert_equal(add_strings('15', '1'), 16)
assert_equal(add_strings('21', '12'), 33)

File: math/add_strings.py
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        i, j = len(num1) - 1, len(num2) - 1
        result, carry = '', 0

        while i >= 0 or j >= 0 or carry != 0:
            d1 = ord(num1[i]) - ord('0') if i >= 0 else 0
            d2 = ord(num2[j]) - ord('0') if j >= 0 else 0

            carry, digit = divmod(d1 + d2 + carry, 10)
            result = str(digit) + result
            i -= 1; j -= 1

        return result


# 415. Add Strings
# https://leetcode.com/problems/add-strings/description/
